pandas/formats/format.py

```diff
@@ -1951,7 +1951,12 @@ def _format_hierarchical_rows(self):
                 for spans, levels, labels in zip(level_lengths,
                                                  self.df.index.levels,
                                                  self.df.index.labels):
-                    values = levels.take(labels)
+
+                    if levels._can_hold_na:
+                        values = levels.take(labels, fill_value=True)
```

Comment:
If I remove the conditional, the tests `test_to_excel_multiindex`, `test_to_excel_multiindex_cols` and `test_to_excel_multiindex_dates` fail with equivalent tracebacks and the same Exception. Here's the traceback of `test_to_excel_multiindex` as an example:

    ======================================================================
    ERROR: test_to_excel_multiindex (pandas.io.tests.test_excel.Openpyxl20Tests)
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "/home/mpuels/progs/pandas-mpuels/pandas/io/tests/test_excel.py", line 1776, in wrapped
        orig_method(self, *args, **kwargs)
      File "/home/mpuels/progs/pandas-mpuels/pandas/io/tests/test_excel.py", line 1321, in test_to_excel_multiindex
        frame.to_excel(path, 'test1', header=False)
      File "/home/mpuels/progs/pandas-mpuels/pandas/core/frame.py", line 1431, in to_excel
        startrow=startrow, startcol=startcol)
      File "/home/mpuels/progs/pandas-mpuels/pandas/io/excel.py", line 875, in write_cells
        for cell in cells:
      File "/home/mpuels/progs/pandas-mpuels/pandas/formats/format.py", line 1984, in get_formatted_cells
        self._format_body()):
      File "/home/mpuels/progs/pandas-mpuels/pandas/formats/format.py", line 1955, in _format_hierarchical_rows
        values = levels.take(labels, fill_value=True)
      File "/home/mpuels/progs/pandas-mpuels/pandas/indexes/base.py", line 1438, in take
        raise ValueError(msg.format(self.__class__.__name__))
    ValueError: Unable to fill values because Int64Index cannot contain NA

I added the conditional, because `take(level, fill_value=True)` only works when the corresponding level of the MultiIndex contains NaNs. When it doesn't, the aforementioned exception is raised.

Here is a small example:

    df = (pd.DataFrame({'c1': [1,1,2,2],
                        'c2': [None] + "b c d".split(),
                        'v' : [6,7,8,9]})
            .set_index(['c1', 'c2']))

    df

yields

    c1  c2  v
    1       6
    1   b   7
    2   c   8
    2   d   9

    df.index

yields

    MultiIndex(levels=[[1, 2], [u'b', u'c', u'd']],
               labels=[[0, 0, 1, 1], [-1, 0, 1, 2]],
               names=[u'c1', u'c2'])

The first level doesn't contain NaNs, so `.take(_, fill_value=True)` raises an Exception:

    levels_0 = df.index.levels[0]
    labels_0 = df.index.labels[0]
    values_0 = levels_0.take(labels_0, fill_value=True)
    values_0

    ---------------------------------------------------------------------------
    ValueError                                Traceback (most recent call last)
    <ipython-input-16-bbcc31a032be> in <module>()
          1 levels_0 = df.index.levels[0]
          2 labels_0 = df.index.labels[0]
    ----> 3 values_0 = levels_0.take(labels_0, fill_value=True); values_0

    /home/mpuels/progs/pandas-mpuels/pandas/indexes/base.pyc in take(self, indices, axis, allow_fill, fill_value, **kwargs)
       1436             if allow_fill and fill_value is not None:
       1437                 msg = 'Unable to fill values because {0} cannot contain NA'
    -> 1438                 raise ValueError(msg.format(self.__class__.__name__))
       1439             taken = self.values.take(indices)
       1440         return self._shallow_copy(taken)

    ValueError: Unable to fill values because Int64Index cannot contain NA

If the level of the MultiIndex contains NaNs, `take(_, fill_values=True)` works:

    levels_1 = df.index.levels[1]
    labels_1 = df.index.labels[1]
    values_1 = levels_1.take(labels_1, fill_value=True)
    values_1

    Index([nan, u'b', u'c', u'd'], dtype='object', name=u'c2')